cpp/src/Slice/MetadataValidator.h

```diff
@@ -0,0 +1,104 @@
+        OptionalTextArgument,
+    };
+
+    /// Typedef for functions which can provide additional validation to certain metadata directives.
+    ///
+    /// They take a reference to the instance of metadata we're validating, and a reference to what it was applied on.
+    /// They should return `nullopt` to signal that the metadata was valid. If the metadata is invalid however, they
+    /// should return a string describing the error. This string will be supplied to the user as part of a diagnostic.
+    using ValidationFunc = std::function<std::optional<std::string>(const MetadataPtr&, const SyntaxTreeBasePtr&)>;
+
+    struct MetadataInfo
+    {
+        /// A list of types that this metadata can validly be applied to.
+        ///
+        /// If this list is empty we don't perform any automatic checking of whether this metadata is validly applied.
+        /// Usually, this is because determining validity isn't as straightforward as matching against a list,
+        /// and requires a more complex approach, which is achieved through providing an `extraValidation` function.
+        std::list<const std::type_info*> validOn;
+
+        /// Specifies how many, and what kinds of arguments, this metadata accepts.
+        MetadataArgumentKind acceptedArguments;
+
+        /// This field stores the specific values that can be provided as arguments to this metadata.
+        /// If this field is unset, then we perform no validation of the arguments (ie. arguments can have any value).
+        std::optional<StringList> validArgumentValues = std::nullopt;
+
+        /// Indicates whether this metadata can be applied to definitions and declarations.
+        bool isDefinitionMetadata = true;
+
+        /// Indicates whether this metadata can be applied to types (ex: on return types, inner sequence types, etc.).
+        ///
+        /// Note that there is special logic in the validator for when this field is `true` and `isDefinitionMetadata`
+        /// is `false`. If these conditions are met and we're validating metadata that has been applied to an operation,
+        /// we validate that metadata for the operation's **return type**, instead of the operation itself.
+        /// We do the same thing for metadata applied to parameters and data members as well.
+        /// Due to the syntax of Slice, it's ambiguous whether metadata is applied to these elements or their types.
+        bool isTypeMetadata = false;
+
+        /// Indicates whether it's valid and meaningful for this metadata to appear multiple times on the same thing.
+        bool mustBeUnique = true;
+
+        /// A function used to run additional validation for this metadata.
+        /// If this field is set, it will always be run.
+        /// @see ValidationFunc
+        ValidationFunc extraValidation = nullptr;
+    };
+
+    class MetadataValidator final : public ParserVisitor
```

Comment:
This is not the correct API for validation. This visitor is an implementation detail that should not be exposed to "clients" (slice2cpp, slice2java).

A much better API is a simple validateMetadata function on Unit:
```cpp
// That's what slice2cpp calls:
unit->validateMetadata(...);
```

Then this MetadataValidator is used in the _implementation_ of `Unit::validateMetadata`.

Reply:
What I ended up doing was making the visitor inheritance private. It is an implementation detail, and indeed shouldn't be exposed. In it's place I added a single "entry-point" function, but instead of `Unit::validateMetadata(MetadataValidator)`, I added `MetadataValidator::validateMetadataWithin(Unit)`.

For two reasons:
1) "Calling a function on a validator, and passing in the thing to validate" feels more natural than
"calling a function on the thing you want to validate, and passing in the validator".
2) The includes are more logical this way. Currently, `MetadataValidator.h` includes `Parser.h`, and that's it (to get all the grammar types). Implementing your suggestions means that we'd have a circular includes (since `Unit` lives in `Parser.h` and now needs to take a `MetadataValidator` parameter). There's ways to address this, but I think keeping the including in a single direction is nicer.